doc/decisions/1_in_discussion/reference_counting.md

```diff
@@ -0,0 +1,286 @@
+# Reference Counting
+
+## Problem
+
+Elektra's data structures are shared between multiple owners.
+A `Key` can be owned by many different `KeySet`s.
+Both `Key` and `KeySet` can be used in bindings.
+If the languages of those bindings have automatic memory management, this also implies a shared ownership.
+
+To solve this issue of shared ownership, we need reference counting.
+
+An adjacent problem is that the names of `Key`s must be read-only while the `Key` is part of a `KeySet`.
+This can also be solved by a secondary counter just for `KeySet`s.
+
+There has also been [discussion](copy_on_write.md) about adding further data structures, which would need reference counters as well.
+
+## Constraints
+
+- Deleting data structures that still have references must be prevented
```

Comment:
```suggestion
- Deleting (parts of) data structures that still have references must be prevented. (Valgrind must not detect any problems.)
```

Reply:
I don't like the Valgrind part... Yes, it's a simple test, but if someone finds an issue that is not detected by valgrind, that's still a problem.

Reply:
Valgrind will detect any problem if there is a test case for this situation. I updated:

```suggestion
- Deleting (parts of) data structures that still have references must be prevented. (Valgrind must not detect any problems for any test case.)
```

Reply:
Still don't like it. I wasn't suggesting that valgrind might miss a memory problem (although that is possible too, e.g. when `mmap` is used). I was talking about exactly the case that no test case exists. So I suggest

```suggestion
- Deleting (parts of) data structures that still have references must be prevented. This includes, but is not limited to, running all (compatible) tests via `valgrind` without issues.
```

My issue was that just adding the `valgrind` part made it sound like just running `valgrind` without problems would be enough to fulfil the constraint. If you really want to mention `valgrind`, we need to clarify that the constraint goes beyond that.

Reply:
Please simply change the underlying file(s).